cmd/oras/discover.go

```diff
@@ -0,0 +1,179 @@
+			return runDiscover(opts)
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.artifactType, "artifact-type", "", "", "artifact type")
+	cmd.Flags().StringVarP(&opts.outputType, "output", "o", "table", fmt.Sprintf("Format in which to display references (%s, %s, or %s). tree format will show all references including nested", "table", "json", "tree"))
+	option.ApplyFlags(&opts, cmd.Flags())
+	return cmd
+}
+
+func runDiscover(opts discoverOptions) error {
+	ctx, _ := opts.SetLoggerLevel()
+	repo, err := opts.NewRepository(opts.targetRef, opts.Common)
+	if err != nil {
+		return err
+	}
+
+	// discover artifacts
+	root := tree.New(opts.targetRef)
+	desc, err := repo.Resolve(ctx, repo.Reference.Reference)
+	if err != nil {
+		return err
+	}
+	desc, refs, err := getAllReferences(ctx, repo, desc, opts.artifactType, root, opts.outputType == "tree")
+	if err != nil {
+		return err
+	}
+
+	switch opts.outputType {
+	case "tree":
+		tree.Print(root)
+	case "json":
+		printDiscoveredReferencesJSON(desc, *refs)
+	default:
+		fmt.Println("Discovered", len(*refs), "artifacts referencing", opts.targetRef)
+		fmt.Println("Digest:", desc.Digest)
+
+		if len(*refs) != 0 {
+			fmt.Println()
+			printDiscoveredReferencesTable(*refs, opts.Verbose)
+		}
+	}
+	return nil
+}
+
+func getAllReferences(ctx context.Context, repo *remote.Repository, desc ocispec.Descriptor, artifactType string, node *tree.Node, queryGraph bool) (ocispec.Descriptor, *[]artifactspec.Descriptor, error) {
+	var results []artifactspec.Descriptor
+	err := repo.Referrers(ctx, desc, func(referrers []artifactspec.Descriptor) error {
+		for _, r := range referrers {
+			if artifactType == "" || artifactType == r.ArtifactType {
+				results = append(results, r)
+				if !queryGraph {
+					continue
+				}
+				// Find all referrers
+				referrerNode := node.AddPath(r.ArtifactType, r.Digest)
+				_, nestedReferrers, err := getAllReferences(
+					ctx, repo,
+					ocispec.Descriptor{
+						Digest:    r.Digest,
+						Size:      r.Size,
+						MediaType: r.MediaType,
+					},
+					artifactType, referrerNode, queryGraph)
+				if err != nil {
+					return err
+				}
+				results = append(results, *nestedReferrers...)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+	return desc, &results, nil
+}
+
+func printDiscoveredReferencesTable(refs []artifactspec.Descriptor, verbose bool) {
+	typeNameTitle := "Artifact Type"
+	typeNameLength := len(typeNameTitle)
+	for _, ref := range refs {
+		if length := len(ref.ArtifactType); length > typeNameLength {
+			typeNameLength = length
+		}
+	}
+
+	print := func(key string, value interface{}) {
+		fmt.Println(key, strings.Repeat(" ", typeNameLength-len(key)+1), value)
+	}
+
+	print(typeNameTitle, "Digest")
+	for _, ref := range refs {
+		print(ref.ArtifactType, ref.Digest)
+		if verbose {
+			printJSON(ref)
+		}
+	}
+}
+
+func printDiscoveredReferencesJSON(desc ocispec.Descriptor, refs []artifactspec.Descriptor) {
```

Comment:
`printDiscoveredReferencesJSON` should have the same output as the [referrer API](https://github.com/oras-project/artifacts-spec/blob/main/manifest-referrers-api.md).